### PR TITLE
Improvements arm analysis ##anal

### PR DESCRIPTION
**Pull request by alvarofe**
Below an example from a library I am working on
Before

```
|           ; CODE XREF from sym.********(0x9b56)                                                                                                                                                                                                                                                                       
|           0x00014afc      7847c046       uxtab16mi r4, r0, r8, ror 8                                                                                                                                                                                                                                                        
|           0x00014b00      00c09fe5       ldr ip, [pc]                ; [0x14b08:4]=0xfffeec5c                                                                                                                                                                                                                               
|           0x00014b04      0ff08ce0       add pc, ip, pc                                                                                                                                                                                                                                                                     
|           ; DATA XREF from loc.00014afc (0x14b00)                                                                                                                                                                                                                                                                           
\           0x00014b08      5cecfeff       invalid                                                                                                                                                                                                                                                                            
|           ; CODE XREF from sym.******** (0x9bb2)                                                                                                                                                                                                                                                                   
|           0x00014b0c      7847c046       uxtab16mi r4, r0, r8, ror 8                                                                                                                                                                                                                                                        
|           0x00014b10      00c09fe5       ldr ip, [pc]                ; [0x14b18:4]=0xfffeecc4                                                                                                                                                                                                                               
|           0x00014b14      0ff08ce0       add pc, ip, pc                                                                                                                                                                                                                                                                     
|           ; DATA XREF from sym.********(0x14b10)                                                                                                                                                                                                                                                                  
\           0x00014b18      c4ecfeff       invalid                                                                                                                                                                                                                                                                            

```
After
```
 loc.00014afc ();                                                                                                                                                                                                                                                                                                          
|           ; CODE XREF from sym.******* (0x9b56)                                                                                                                                                                                                                                                                       
\           0x00014afc      7847           bx pc                                                                                                                                                                                                                                                                              
            0x00014afe      c046           mov r8, r8                                                                                                                                                                                                                                                                         
            0x00014b00      00c09fe5       ldr ip, [pc]                ; [0x14b08:4]=0xfffeec5c                                                                                                                                                                                                                               
            0x00014b04      0ff08ce0       add pc, ip, pc                                                                                                                                                                                                                                                                     
            ; DATA XREF from loc.00014afc (+0x4)                                                                                                                                                                                                                                                                              
            0x00014b08      5cecfeff       invalid                                                                                                                                                                                                                                                                            
|           ; CODE XREF from sym.******* (0x9bb2)                                                                                                                                                                                                                                                                   
\           0x00014b0c      7847           bx pc                                                                                                                                                                                                                                                                              
            0x00014b0e      c046           mov r8, r8                                                                                                                                                                                                                                                                         
            0x00014b10      00c09fe5       ldr ip, [pc]                ; [0x14b18:4]=0xfffeecc4                                                                                                                                                                                                                               
            0x00014b14      0ff08ce0       add pc, ip, pc                                                                                                                                                                                                                                                                     
            ; DATA XREF from loc.00014afc (+0x14)                                                                                                                                                                                                                                                                             
            0x00014b18      c4ecfeff       invalid             
```

This brings slow down during analysis since the look up time for anal hints is incresed but on the other hand is more accurate.

### CODE REVIEW

File: libr/anal/p/anal_arm_cs.c [hunk 1]
```diff
@@ -1477,8 +1477,15 @@ static int analop_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int len
 		break;
 	case ARM_INS_BX:
 	case ARM_INS_BXJ:
-		r_strbuf_setf (&op->esil, "%s,pc,=", ARG(0));
+		{
+		const char *op1 = ARG(0);
```
Comment (radare) [inline libr/anal/p/anal_arm_cs.c]: Missing space

File: libr/anal/p/anal_arm_cs.c [hunk 2]
```diff
@@ -1477,8 +1477,15 @@ static int analop_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int len
...
+		if (!strcmp (op1, "pc")) {
+			r_strbuf_setf (&op->esil, "%d,$$,+,pc,=", pcdelta);
+		} else {
+			r_strbuf_setf (&op->esil, "%s,pc,=", ARG(0));
```
Comment (radare) [inline libr/anal/p/anal_arm_cs.c]: Missing space

File: libr/core/cmd_anal.c [hunk 3]
```diff
@@ -7151,6 +7151,7 @@ static int cmd_anal_all(RCore *core, const char *input) {
 			r_cons_break_push (NULL, NULL);
 			r_cons_break_timeout (r_config_get_i (core->config, "anal.timeout"));
 			r_core_anal_all (core);
+			r_anal_build_range_on_hints(core->anal, true);
```
Comment (radare) [inline libr/core/cmd_anal.c]: Space

**Comment by alvarofe** [comment]
I cannot reproduce the tests failing. As example the fatmach0 which is marked XX in travis, locally is running fine. Should I run test differently to reproduce?
```
ryzen:new alvaro$ node bin/r2r.js db/formats/mach0/fatmach0 
[**]      db/formats/mach0/fatmach0   OK   23 BR    2 XX    0 FX    0

ryzen:radare2 alvaro$ r2 -version
radare2 3.1.0-git 19969 @ linux-x86-64 git.2.9.0-586-gd91a56b76
commit: d91a56b76e8a9ced8a55d8150e028fe4da04de95 build: 2018-11-07__11:22:33

ryzen:radare2 alvaro$ git log                                                                                                                                                                                                                                                                     
commit d91a56b76e8a9ced8a55d8150e028fe4da04de95 (HEAD -> fix)                                                                                                                                                                                                                                                                 
Author: Alvaro <alvaro.felipe91@gmail.com>                                                                                                                                                                                                                                                                                    
Date:   Tue Nov 6 18:33:15 2018 +0100                                                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                              
    clean anal hints after aa 
```

File: libr/core/cmd_anal.c (hunk 3 above)
Comment (radare) [reply libr/core/cmd_anal.c]: Indent

**Comment by radare** [comment]
Fails because of a timeout. Travis runs for half an hour and doesnt reaches the end 

> On 7 Nov 2018, at 11:28, Álvaro Felipe Melchor <notifications@github.com> wrote:
> 
> I cannot reproduce the tests failing. As example the fatmach0 which is marked XX in travis, locally is running fine. Should I run test differently to reproduce?
> 
> ryzen:new alvaro$ node bin/r2r.js db/formats/mach0/fatmach0 
> [**]      db/formats/mach0/fatmach0   OK   23 BR    2 XX    0 FX    0
> 
> ryzen:radare2 alvaro$ r2 -version
> radare2 3.1.0-git 19969 @ linux-x86-64 git.2.9.0-586-gd91a56b76
> commit: d91a56b76e8a9ced8a55d8150e028fe4da04de95 build: 2018-11-07__11:22:33
> 
> ryzen:radare2 alvaro$ git log                                                                                                                                                                                                                                                                     
> commit d91a56b76e8a9ced8a55d8150e028fe4da04de95 (HEAD -> fix)                                                                                                                                                                                                                                                                 
> Author: Alvaro <alvaro.felipe91@gmail.com>                                                                                                                                                                                                                                                                                    
> Date:   Tue Nov 6 18:33:15 2018 +0100                                                                                                                                                                                                                                                                                         
>                                                                                                                                                                                                                                                                                                                               
>     clean anal hints after aa 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by alvarofe** [comment]
I was referring to this, which is not due to timeout
```
[XX] db/formats/mach0/fatmach0 fatmach0 sections  106
$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -a x86 -b 64 -i /tmp/tmp-14761dxm3znrDMJL2.tmp ../bins/mach0/fatmach0-3true
S=~?
-10
EXPECT64=
  1810[**]      db/formats/mach0/fatmach0   OK  993 BR   19 XX    2 FX    1
```

**Comment by alvarofe** [comment]
About the performance I will look into it with the aim of improving the times.

**Comment by radare** [comment]
now the timeout is gone, but there are many XX still, do you have a branch in r2r or you missed to rebase?

File: libr/anal/anal.c [hunk 4]
```diff
@@ -618,45 +643,77 @@ R_API bool r_anal_noreturn_at(RAnal *anal, ut64 addr) {
...
+				range->from = addr;
+				r_anal_range_hint_tree_insert (&a->rb_hints_ranges, range);
 			}
+			range_bits = bits;
 		}
-		ls_free (sdb_range);
-		a->bits_hints_changed = false;
+		a->merge_hints = false;
+	}
+
```
Comment (radare) [inline libr/anal/anal.c]: empty line

**Comment by alvarofe** [comment]
Dunno why there are some XX, I don't get the same result locally

```
ryzen:new alvaro$ node bin/r2r.js db/formats/mach0/fatmach0                                                                                                                                                                                                                                                                  
[**]      db/formats/mach0/fatmach0   OK   23 BR    2 XX    0 FX    0
ryzen:new alvaro$ node bin/r2r.js db/esil/apple
[**]                  db/esil/apple   OK    8 BR    0 XX    0 FX    0


ryzen:radare2 alvaro$ r2 -version
radare2 3.1.0-git 20204 @ linux-x86-64 git.3.0.1-218-g8d053c470
commit: 8d053c470fc9989ac405755845c0a59dcf1a38a6 build: 2018-11-13__18:52:07

ryzen:radare2 alvaro$ git log  
commit 8d053c470fc9989ac405755845c0a59dcf1a38a6 (HEAD -> fix, origin/fix)                                                                                                                                                                                                                                                     
Author: Alvaro <alvaro.felipe91@gmail.com>                                                                                                                                                                                                                                                                                    
Date:   Tue Nov 6 18:33:15 2018 +0100                                                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                              
    anal: arm: handle better anal hints to incrase performance ##anal            
```

However, in travis

```
[XX] db/esil/apple sectonly esil stack  77
$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -i /tmp/tmp-15759Z9fhFZSNpPJB.tmp ../bins/mach0/FileDP
s main
aei
aeim
e asm.emu=true
e asm.emustr=true
e io.sectonly=false
.ar*
pxW 4 @ sp
e io.sectonly=true
pxW 4 @ sp
S 0 0x100000 0 983040 stack rw- 0
pxW 4 @ sp
 0x00178000 0x00000000 sb
 0x00178000 0x00000000 sb
-0x00178000 0x00000000 sb

[XX] db/formats/mach0/fatmach0 fatmach0 sections  120
$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -a x86 -b 64 -i /tmp/tmp-157597z6d57hgelvd.tmp ../bins/mach0/fatmach0-3true
S=~?
-10
```

Should I run test differently? cc @Maijin

File: libr/anal/hint.c [hunk 5]
```diff
@@ -111,8 +109,11 @@ R_API void r_anal_hint_set_esil(RAnal *a, ut64 addr, const char *esil) {
 }
 
 R_API void r_anal_hint_set_bits(RAnal *a, ut64 addr, int bits) {
-	a->bits_hints_changed = true;
 	setHint (a, "bits:", addr, NULL, bits);
+	if (a && a->hint_cbs.on_bits) {
+		a->hint_cbs.on_bits(a, addr, bits, true);
```
Comment (radare) [inline libr/anal/hint.c]: missing space

**Comment by radare** [comment]
any reason to use rbtree? its only used in fcn.c and im in th eprocess of moving all the daa structures that we use in 1 place only into sdb/Attic for later discussion, testing, evaluation and re-introduction into r2 via sdb

**Comment by alvarofe** [comment]
The reason is the lookups time performances against rlist. Easily you might end up with 4K anal hints or more with the new logic, rbtree reduces the time when searching for the right hint and thus the right bits

File: libr/anal/anal.c [hunk 6]
```diff
@@ -55,6 +55,45 @@ static void zign_rename_for(void *user, int idx, const char *oname, const char *
...
+							   RAnalRange, rb);
+			if (range1->rb_max_addr > range->rb_max_addr) {
+				range->rb_max_addr = range1->rb_max_addr;
+			}
+		}
+	}
+}
+
+static int __anal_hint_range_tree_cmp(const void *a_, const RBNode *b_) {
+	const RAnalRange *a = (const RAnalRange *)a_;
```
Comment (lzutao) [inline libr/anal/anal.c]: Does it need to cast void *? In above function you don't cast but you cast it here.

**Comment by radare** [comment]
Please resolve the conflicts and rebase

File: libr/anal/anal.c [hunk 7]
```diff
@@ -55,6 +55,45 @@ static void zign_rename_for(void *user, int idx, const char *oname, const char *
...
+	const RAnalRange *b = (const RAnalRange *)container_of (b_, const RAnalRange, rb);
+	if (a && b) {
+		ut64 from0 = a->from, from1 = b->from;
+		return from0 < from1 ? -1 : 1;
+	}
+	return 0;
+}
+
+static void __anal_hint_range_tree_free(RBNode *node) {
+	free (container_of(node, RAnalRange, rb));
```
Comment (radare) [inline libr/anal/anal.c]: missing space

File: libr/core/disasm.c [hunk 8]
```diff
@@ -4749,9 +4749,8 @@ R_API int r_core_print_disasm(RPrint *p, RCore *core, ut64 addr, ut8 *buf, int l
 	if (!ds->l) {
 		ds->l = core->blocksize;
 	}
-
+	r_anal_merge_hint_ranges(core->anal);
```
Comment (radare) [inline libr/core/disasm.c]: missing spce

File: libr/anal/hint.c [hunk 9]
```diff
@@ -124,8 +125,11 @@ R_API void r_anal_hint_unset_size(RAnal *a, ut64 addr) {
 }
 
 R_API void r_anal_hint_unset_bits(RAnal *a, ut64 addr) {
-	a->bits_hints_changed = true;
 	unsetHint(a, "bits:", addr);
+	if (a && a->hint_cbs.on_bits) {
+		a->hint_cbs.on_bits(a, addr, 0, false);
```
Comment (radare) [inline libr/anal/hint.c]: space

File: libr/anal/anal.c [hunk 10]
```diff
@@ -618,45 +643,77 @@ R_API bool r_anal_noreturn_at(RAnal *anal, ut64 addr) {
...
-		ls_free (sdb_range);
-		a->bits_hints_changed = false;
+		a->merge_hints = false;
+	}
+
+}
+
+
+R_API void r_anal_build_range_on_hints(RAnal *a, ut64 addr, int bits) {
+	if (!a) {
```
[the remark this answers is not in the document]
Comment (radare) [reply libr/anal/anal.c]: i think this function name is wrong

[the remark this answers is not in the document]
Comment (alvarofe) [reply libr/anal/anal.c]: the name is the same as before, propose a better one and i will change it

**Comment by radare** [comment]
ok

**Comment by radare** [comment]
yes, but you changed the implementation and its doing a different thing now, there's no "hints changed" bit because its not regenerating any cache here (im with coffee, so maybe i missed something)

**Comment by alvarofe** [comment]
you are right, now is adding new hints as mean they are created. You should call merge explicitly to clean up the ranges when the analisis has finished. Is performed when printing the disassembly. Asking for a name because I am not very good at `r_anal_add_range` or something alike?

File: libr/anal/anal.c (hunk 6 above)
Comment (alvarofe) [reply libr/anal/anal.c]: fixed

**Comment by radare** [comment]
cc @alvarofe remove the 'fix' branch in your r2r. and fix those 3 tests after rebasing

https://api.travis-ci.org/v3/job/456768489/log.txt

**Comment by radare** [comment]
One of the tests can be fixed with this:

screenshot 2018-11-21 at 23 50 52

this is... if you do 'pd' before 'aa' the analysis of the function is correct, otherwise its not.

The reason for that is because when you run pd, the hints are resetted, the ones coming from rbin are gone and new ones appear. this cant be fixed with ah-*, because this way it fails with other errors. so i guess the answer is in disasm.c,

**Comment by alvarofe** [comment]
The two test which are [XX] should be marked as BR. I already opened an issue to track those tests https://github.com/radare/radare2/issues/12146 - at the moment -m option with binaries does not work, since it does not pick up the right bin plugin, hints are not set and thus, the analysis fail.

**Comment by radare** [comment]
lgtm, please send a PR in r2r marking them as broken or update the output

**Comment by radare** [comment]
Pls @maijin take a look at these tests checking the output instead of the numebrs

Good to see the other two xx being green now

**Comment by Maijin** [comment]
Not sure you need me for doing this but here you go:

Same difference in both tests:

## raw aac with maps (using a PIC bin) 

```
< 0x000080ac    6 64           fcn.000080ac
< 0x000080f4    1 4            fcn.000080f4
---
> 0x000080ac    6 50           fcn.000080ac
> 0x000080de    1 22           fcn.000080de
> 0x000080f4    1 8            fcn.000080f4
165c166
< 0x00012cd0    1 16           fcn.00012cd0
---
> 0x00012cd0    1 2            fcn.00012cd0
170c171
< 0x00012d20    1 16           fcn.00012d20
---
> 0x00012d20    1 2            fcn.00012d20
185c186
< 0x00012e10    1 16           fcn.00012e10
---
> 0x00012e10    1 2            fcn.00012e10
187c188
< 0x00012e30    1 16           fcn.00012e30
---
> 0x00012e30    1 2            fcn.00012e30
189c190
< 0x00012e50    1 16           fcn.00012e50
---
> 0x00012e50    1 2            fcn.00012e50
```

## raw aac (using a PIC bin)
```
< 0x000080ac    6 64           fcn.000080ac
< 0x000080f4    1 4            fcn.000080f4
---
> 0x000080ac    6 50           fcn.000080ac
> 0x000080de    1 22           fcn.000080de
> 0x000080f4    1 8            fcn.000080f4
165c166
< 0x00012cd0    1 16           fcn.00012cd0
---
> 0x00012cd0    1 2            fcn.00012cd0
170c171
< 0x00012d20    1 16           fcn.00012d20
---
> 0x00012d20    1 2            fcn.00012d20
185c186
< 0x00012e10    1 16           fcn.00012e10
---
> 0x00012e10    1 2            fcn.00012e10
187c188
< 0x00012e30    1 16           fcn.00012e30
---
> 0x00012e30    1 2            fcn.00012e30
189c190
< 0x00012e50    1 16           fcn.00012e50
---
> 0x00012e50    1 2            fcn.00012e50
```

I took one example:

### Before patch 

```
[0x00005fa8]> pdf @ fcn.00012cd0
┌ (fcn) fcn.00012cd0 16
│   fcn.00012cd0 ();
│           ; CALL XREFS from sym.decode__JNIEnv___jstring___jstring (0x668e, 0x669a)
│           0x00012cd0      7847c046       uxtab16mi r4, r0, r8, ror 8
│           0x00012cd4      00c09fe5       ldr ip, [pc]                ; [0x12cdc:4]=0xffff3178
│           0x00012cd8      0ff08ce0       add pc, ip, pc
│           ; DATA XREF from fcn.00012cd0 (0x12cd4)
└           0x00012cdc      7831ffff       invalid
```

### After patch

```
[0x00005fa8]> pdf @ fcn.00012cd0
┌ (fcn) fcn.00012cd0 2
│   fcn.00012cd0 ();
│           ; CALL XREFS from sym.decode__JNIEnv___jstring___jstring (0x668e, 0x669a)
└           0x00012cd0      7847           bx pc
```

**Pull request merged**